llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp

```diff
@@ -6989,9 +6989,14 @@ OpenMPIRBuilder::InsertPointTy OpenMPIRBuilder::emitTargetTask(
         getOrCreateSrcLocStr(LocationDescription(Builder), SrcLocStrSize);
     Value *Ident = getOrCreateIdent(SrcLocStr, SrcLocStrSize);
 
-    // @__kmpc_omp_task_alloc
+    // @__kmpc_omp_task_alloc or @__kmpc_omp_target_task_alloc
+    //
+    // If `HasNoWait == true`, we call  @__kmpc_omp_target_task_alloc to provide
+    // the DeviceID to the deferred task.
```

Comment:
IIUC, we are also interested in the async behavior of `@__kmpc_omp_target_task_alloc`, that sets the `TASK_UNTIED` flag for this, right? A brief comment about this here could help to understand why `@__kmpc_omp_target_task_alloc` is needed when `HasNoWait == true`.

Interestingly enough, the current implementation of `@__kmpc_omp_target_task_alloc` doesn't even use the DeviceID argument.

Reply:
OpenMP API says:

![image](https://github.com/user-attachments/assets/24d623d3-b8d6-4a4a-be22-e714b76fe501)

So, the `united` flag should be set when the target task is generated.

Reply:
Looking at the OpenMP runtime code, handling `untied` is in the runtime in the `__kmpc_omp_target_task_alloc` entry point.